common/trinary/trit_array.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled/
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include "common/trinary/trit_array.h"
+#include <stdlib.h>
+#include <string.h>
+#include "common/trinary/trit_byte.h"
+
+size_t flex_trit_array_slice(flex_trit_t *to_trit_array, size_t to_len,
+                             const flex_trit_t *trit_array, size_t len,
+                             size_t start, size_t num_trits) {
+  // Bounds checking
+  if (num_trits > to_len || (start + num_trits) > len) {
+    return 0;
+  }
+  size_t num_bytes = trit_array_bytes_for_trits(num_trits);
+#if defined(TRIT_ARRAY_ENCODING_1_TRIT_PER_BYTE)
+  memcpy(to_trit_array, trit_array + start, num_trits);
+#elif defined(TRIT_ARRAY_ENCODING_4_TRITS_PER_BYTE)
+  uint8_t buffer;
+  uint8_t tshift = (start & 3) << 1U;
+  uint8_t rshift = (8U - tshift) % 8U;
+  size_t index = start >> 2U;
+  size_t end_index = (start + num_trits - 1) >> 2U;
+  size_t i, j;
+  // Calculate the number of bytes to copy over
+  for (i = index, j = 0; i < index + num_bytes; i++, j++) {
+    buffer = trit_array[i];
+    buffer = buffer >> tshift;
+    if (i < end_index) {
+      if (rshift) {
+        buffer |= (trit_array[i + 1] << rshift);
```

Comment:
aren't last two shifts the opposite?

Reply:
tshift != rshift though?

Reply:
Yes, rshift and shift are different and they go in the opposite direction, in case we need to shift some trits from the next byte to the buffer